### PR TITLE
Persist referral link

### LINKED ISSUE #850: [Affiliate] Persist the affiliate link in the browser
**Description**
Persist the affiliate link in the browser

**Proposal**
- Persist in the global state of the app, associated with the connected address, remember the user can change addresses any time
- ~Make sure when the user loads the app we load from the state the IPFS hash if available~

**Additional context**
Draft wireframes proposal https://www.figma.com/file/P0AIwKkIUpEZtLukkCUM94/CowSwap_-Affiliate?node-id=48%3A224
Affiliate main doc https://docs.google.com/document/d/1PsdWn9WPFf2BNeoJNtClyra8jNvWk_bJ-g33TbpUf8s/edit#heading=h.wc3ppxg7l1mo

## Update 2021/07/19

Removed reference to IPFS and it has been de-scoped for now

### PR DESCRIPTION
**Pull request by stackseeder**
# Summary

Fixes #850 

When we use the referral link, referrer address is persisted in the state.

No visible change here. (No screenshot)

  # To Test

1. Copy referral link (or emulate any referral link like http(s)://<domain>/swap?referral=<address>)
- [] Paste url in another tab.
- [] Referrer address will be stored in the local storage like {"referrerAddress":"<Address>"}

### CODE REVIEW

File: src/components/Web3ReactManager/index.tsx [hunk 1]
```diff
@@ -48,6 +53,20 @@ export default function Web3ReactManager({ children }: { children: JSX.Element }
...
+
+  const startWalletListeners = async () => {
+    const provider: any = await detectEthereumProvider()
+    if (!provider && active) return
+    provider.on('accountsChanged', function (accounts: Array<string>) {
+      console.log('ACCOUNT CHANGED', accounts)
+      generateAffiliateLink((link: string) => dispatch(updateAffiliateLink({ affiliateLink: link })))
+    })
+  }
+
```
Comment (ramirotw) [inline src/components/Web3ReactManager/index.tsx:69]: I'm not sure why this handler is needed to detect an account change. Why not just use the `account` property of `useWalletInfo` in a useEffect?

File: src/hooks/web3.ts [hunk 2]
```diff
@@ -57,17 +69,25 @@ export function useInactiveListener(suppress = false) {
...
-          activate(injected, undefined, true).catch((error) => {
-            console.error('Failed to activate after accounts changed', error)
-          })
+          activate(injected, undefined, true)
+            .catch((error) => {
+              console.error('Failed to activate after accounts changed', error)
+            })
+            .finally(() =>
+              generateAffiliateLink((link: string) => dispatch(updateAffiliateLink({ affiliateLink: link })))
+            )
```
Comment (ramirotw) [inline src/hooks/web3.ts:90]: Is this necessary? Remember that this file is not under the `custom` folder.

File: src/utils/affiliateLinkUtil.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  if (typeof window.ethereum == 'undefined') return
+  const provider: any = await detectEthereumProvider()
+  const accounts: Array<string> = await provider.request({
+    method: 'eth_requestAccounts',
+  })
+  let link = null
+  if (accounts.length > 0) link = `${window.location.href}/#/?referral=${accounts[0]}`
+  console.log('GEN LINK =>', link)
+  if (callback) return callback(link)
+}
```
Comment (ramirotw) [inline src/utils/affiliateLinkUtil.ts:13]: There's already a `useReferralLink` hook to do this.

File: package.json [hunk 4]
```diff
@@ -168,6 +168,7 @@
     "@gnosis.pm/cow-runner-game": "^0.2.9",
     "@gnosis.pm/dex-js": "^0.11.0",
     "@gnosis.pm/gp-v2-contracts": "^0.0.1-alpha.20",
+    "@metamask/detect-provider": "^1.2.0",
```
Comment (ramirotw) [inline package.json:171]: is this library really needed?

**Review by ramirotw — CHANGES_REQUESTED** [review]

**Review by ramirotw — COMMENTED** [review]
Check the build also which has a few errors https://github.com/gnosis/cowswap/pull/1025/checks?check_run_id=3135399538

File: src/hooks/web3.ts (hunk 2 above)
Comment (stackseeder) [reply src/hooks/web3.ts:89]: @ramirotw do you want to move this file to custom folder?
It is used in too many files already.

File: yarn.lock [hunk 5]
```diff
@@ -2300,6 +2300,11 @@
     "@babel/runtime" "^7.11.2"
     "@lingui/core" "^3.10.2"
 
+"@metamask/detect-provider@^1.2.0":
+  version "1.2.0"
+  resolved "https://registry.yarnpkg.com/@metamask/detect-provider/-/detect-provider-1.2.0.tgz#3667a7531f2a682e3c3a43eaf3a1958bdb42a696"
+  integrity sha512-ocA76vt+8D0thgXZ7LxFPyqw3H7988qblgzddTDA6B8a/yU0uKV42QR/DhA+Jh11rJjxW0jKvwb5htA6krNZDQ==
+
```
Comment (ramirotw) [inline yarn.lock:2307]: This dependency can be removed

File: src/hooks/web3.ts [hunk 6]
```diff
@@ -49,25 +62,39 @@ export function useEagerConnect() {
...
-          })
+          activate(injected, undefined, true)
+            .catch((error) => {
+              console.error('Failed to activate after accounts changed', error)
+            })
+            .finally(() => {
+              if (referralLink) {
+                dispatch(updateAffiliateLink({ affiliateLink: referralLink }))
+              }
+            })
```
Comment (ramirotw) [inline src/hooks/web3.ts:97]: I still don't understand why do we need this finally clauses. Can you elaborate on why this file needs to be updated?

File: src/custom/components/Web3ReactManager/index.tsx [hunk 7]
```diff
@@ -48,6 +55,12 @@ export default function Web3ReactManager({ children }: { children: JSX.Element }
     }
   }, [])
 
+  useEffect(() => {
+    if (referralLink) {
+      dispatch(updateAffiliateLink({ affiliateLink: referralLink }))
+    }
+  }, [account, referralLink, dispatch])
+
```
Comment (ramirotw) [inline src/custom/components/Web3ReactManager/index.tsx:63]: We shouldn't be updating either moving this Web3ReactManager file. Create a new hook which does this integration and put it in the App/index.ts file so it runs when the app renders.

**Review by ramirotw — CHANGES_REQUESTED** [review]

File: yarn.lock (hunk 5 above)
Comment (stackseeder) [reply yarn.lock]: I've removed it.

File: src/hooks/web3.ts (hunk 6 above)
Comment (stackseeder) [reply src/hooks/web3.ts]: I did not use useEffect before so we required this code. But now I am using useEffect, so don't need to specify the code in this file. I've removed it.

**Review by alfetopito — CHANGES_REQUESTED** [review]
I'm sorry Ivica, but that's now how our Modding system works.

Please do not move the original files out of their places, not modify the original files unless strictly necessary.
Take a look at this doc for reference https://docs.google.com/document/d/1-4b_BXevDPq1-anyx9qFnocshlUqom5W1lYuA2Y65us/edit#heading=h.lb282ue6cmsr
I'm happy to go over it with you if you have any questions.

File: src/hooks/web3.ts [hunk 8]
```diff
@@ -82,5 +86,5 @@ export function useInactiveListener(suppress = false) {
       }
     }
     return undefined
-  }, [active, error, suppress, activate])
+  }, [active, error, suppress, activate, dispatch])
```
Comment (ramirotw) [inline src/hooks/web3.ts:89]: Changes in this file are outdated

File: src/state/affiliate/reducer.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  affiliateLink: string
+}
+
+export const initialState: AffiliateLinkState = {
+  affiliateLink: '',
+}
+
+export default createReducer(initialState, (builder) =>
+  builder.addCase(updateAffiliateLink, (state, action) => {
+    console.log('Triggered Updated Link')
```
Comment (ramirotw) [inline src/state/affiliate/reducer.ts:14]: nit: Remove this console.log

**Review by ramirotw — CHANGES_REQUESTED** [review]
Looking good!

Just a few things to update

File: src/custom/state/index.ts [hunk 10]
```diff
@@ -19,7 +19,7 @@ import orders from './orders/reducer'
 import price from './price/reducer'
 import gas from 'state/gas/reducer'
 import { updateVersion } from 'state/global/actions'
-import affiliate from '@src/state/affiliate/reducer'
+import affiliate from '@src/custom/state/affiliate/reducer'
```
Comment (ramirotw) [inline src/custom/state/index.ts:22]: remove the `@src/custom/` path prefix

**Review by ramirotw — CHANGES_REQUESTED** [review]
Please remove the remaining changes in the `src/hooks/web3.ts` file

**Review by ramirotw — APPROVED** [review]

File: src/custom/state/affiliate/updater.tsx [hunk 11]
```diff
@@ -0,0 +1,20 @@
+import { useEffect } from 'react'
+
+import { updateAffiliateLink } from '@src/custom/state/affiliate/actions'
+import useReferralLink from '../../hooks/useReferralLink'
+import { useWalletInfo } from '../../hooks/useWalletInfo'
+import { useAppDispatch } from '@src/state/hooks'
+
+export default function ReferralLinkUpdater() {
+  const dispatch = useAppDispatch()
+  const referralLink = useReferralLink()
```
Comment (alfetopito) [inline src/custom/state/affiliate/updater.tsx:10]: Actually, this returns your own affiliate link, the one you want to share with other people.
What we want here is to store what's coming from the url param when someone else opens your URL affiliate link.

You'll want to get the value Ramiro is fetching on this file from the PR https://github.com/gnosis/cowswap/pull/953/files#diff-af06422a6866f6f150604ce5491f449819b0882344b8da8b2387f6288fe74949L1

See the comments there.
The one where you said useless code? That's where you'll trigger your stuff :)

**Review by alfetopito — CHANGES_REQUESTED** [review]
Great job, you the spirit of the mod files now :)

Comments below

**Review by alfetopito — CHANGES_REQUESTED** [review]
The app doesn't work with the latest changes
screenshot_2021-07-28_15-07-25

Please make sure it does before submitting the PR :)

As for what's wrong, I did not debug the issue, but I guess you have to ditch your updater and instead trigger the update from this file somehow https://github.com/gnosis/cowswap/pull/953/files#diff-af06422a6866f6f150604ce5491f449819b0882344b8da8b2387f6288fe74949L1

I assume it's loading the url param too soon.
You can sync with Ramiro who coded that part, and as far I can tell, was working fine.

**Comment by ramirotw** [comment]
> screenshot_2021-07-28_15-07-25
> 
> Please make sure it does before submitting the PR :)
> 
> As for what's wrong, I did not debug the issue, but I guess you have to ditch your updater and instead trigger the update from this file somehow https://github.com/gnosis/cowswap/pull/953/files#diff-af06422a6866f6f150604ce5491f449819b0882344b8da8b2387f6288fe74949L1
> 
> I assume it's loading the url param too soon.
> You can sync with Ramiro who coded that part, and as far I can tell, was working fine.

The problem is that the `Updaters` component is rendered above in the tree than the `HashRouter`, so the location's context is not defined yet. A solution could be to move the Updaters below the HashRouter, or to instantiate the ReferralLinkUpdater as @alfetopito mentioned in the App file (btw, you should also remove the `useParseReferralQueryParam` reference in the `src/custom/pages/App/index.tsx` file which I left just for testing).

**Comment by ramirotw** [comment]
@stackseeder Please fix the build error:

> src/custom/pages/App/index.tsx
> Line 1:17:  'useEffect' is defined but never used  @typescript-eslint/no-unused-vars

File: src/custom/state/affiliate/updater.tsx (hunk 11 above)
Comment (ramirotw) [reply src/custom/state/affiliate/updater.tsx]: > What we want here is to store what's coming from the url param when someone else opens your URL affiliate link.

@alfetopito I believe there's a misunderstanding with the ticket's description:

> Persist in the global state of the app, associated with the connected address, remember the user can change addresses any time

The ticket mentions the connected address while your comment says its the address from the url.

**Review by alfetopito — COMMENTED** [review]
It all looks good now, but doesn't seem to be working.

I'm trying to debug this because I'm not sure whether this change has no effect of the URL loader is not working at all.

I followed the steps defined in the test section, but the `affiliateLink` on local storage is never updated
screenshot_2021-07-29_11-59-09

I've accessed it with a random `referral` https://pr1025--gpswapui.review.gnosisdev.com/#/swap?referral=adsadsa

Still empty on local storage.

Maybe is my setup. Can you try it out as well?

**Comment by stackseeder** [comment]
@alfetopito sorry, the original PR was doing wrong action, so the original description was also wrong.
I've updated it.
I've also changed some names. It was storing referral link, but now storing referrer address, so I've changed the variable name from 'affiliateLink' to 'referrerAddress'.

**Review by alfetopito — APPROVED** [review]
Awesome!

**Pull request merged**